src/llt/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+unsafe extern "system" fn invoker<T>(arg: *mut c_void) -> u32
+where
+    T: FnMut() + Send + 'static,
+{
+    use std::arch::asm;
+    use windows_sys::Win32::System::Memory::VirtualFree;
+    use windows_sys::Win32::System::Threading::ExitThread;
+
+    // We can't keep any variables that need to be dropped on the stack because we need to exit the
+    // thread with ExitThread(). In this case any variables on the stack will not get dropped, which
+    // will cause a memory to leak.
+    let (entry, stack, stack_size) = *Box::from_raw(arg as *mut (T, *mut u8, usize));
+
+    assert!(ENTRY.set(UnsafeCell::new(Entry(Box::new(entry)))).is_none());
+
+    // Switch stack then invoke the entry.
+    unsafe extern "system" fn run() {
+        // All local variables here don't need to be dropped.
+        let entry = ENTRY.get().unwrap();
+        let entry = entry.get();
+
+        (*entry).0();
+    }
+
+    asm!(
+        // Set stack limit.
+        "mov rcx, gs:[0x1478]",
+        "mov gs:[0x10], rax",
+        "mov gs:[0x1478], rax",
+        // Set stack base.
+        "add rax, rdx",
+        "mov rsp, rax",
+        "mov gs:[0x08], rax",
+        // Set SEH frame as the end of frame.
+        "mov rax, 0xFFFFFFFFFFFFFFFF",
+        "mov gs:[0x00], rax",
+        // Set guaranteed bytes to zero.
+        "xor eax, eax",
+        "mov gs:[0x1748], eax",
+        // Free system provided stack.
+        // TODO: Panic if error.
+        "xor rdx, rdx",
+        "mov r8d, 0x8000",
+        "sub rsp, 32",
+        "call {free}",
+        // Run the entry.
+        "call {run}",
+        // Exit the thread.
+        "xor ecx, ecx",
+        "call {exit}",
+        in("rax") stack,
+        in("rdx") stack_size,
+        free = sym VirtualFree,
+        run = sym run,
+        exit = sym ExitThread,
+        options(noreturn)
+    );
+}
+
+static ENTRY: Tls<UnsafeCell<Entry>> = Tls::new();
+
+struct Entry(Box<dyn FnMut()>);
+
+#[cfg(windows)]
+impl Drop for Entry {
+    fn drop(&mut self) {
+        use std::arch::asm;
+
+        // Set DeallocationStack to null to prevent Windows free it. We did not do this in the
+        // invoker because we want SetThreadStackGuarantee to be working with the user-provided stack.
+        unsafe {
+            asm!(
+                "xor rax, rax",
+                "mov gs:[0x1478], rax",
+                out("rax") _
+            )
+        };
+    }
+}
+
+#[cfg(unix)]
+pub type Thread = libc::pthread_t;
+
+#[cfg(windows)]
+pub type Thread = windows_sys::Win32::Foundation::HANDLE;
+
+#[derive(Debug, Error)]
+pub enum SpawnError {
+    #[cfg(unix)]
+    #[error("cannot initialize thread's attribute")]
+    InitAttrFailed(#[source] std::io::Error),
+
+    #[cfg(unix)]
+    #[error("cannot set thread stack")]
+    SetStackFailed(#[source] std::io::Error),
+
+    #[error("cannot create a thread")]
+    CreateThreadFailed(#[source] std::io::Error),
+}
+
+#[cfg(test)]
```

Comment:
If weare going to be adding tests maybe have it checked at CI as well?

Reply:
I was lazy to setup CTest but forgot that we can just run only Rust tests.